lib/milc_interface.cpp

```diff
@@ -1184,6 +1191,895 @@ void qudaEigCGInvert(int external_precision, int quda_precision, double mass, Qu
+    loadGaugeQuda(const_cast<void *>(fatlink), &fat_param);
+    if (longlink != nullptr) loadGaugeQuda(const_cast<void *>(longlink), &long_param);
+    invalidate_quda_gauge = false;
+  }
+
+  mg_pack->mg_preconditioner = newMultigridQuda(&mg_pack->mg_param);
+  mg_pack->last_mass = mass;
+
+  invalidate_quda_mg = false;
+
+  if (!create_quda_gauge) invalidateGaugeQuda();
+
+  qudamilc_called<false>(__func__, verbosity);
+
+  return (void *)mg_pack;
+}
+
+void qudaInvertMG(int external_precision, int quda_precision, double mass, QudaInvertArgs_t inv_args,
+                  double target_residual, double target_fermilab_residual, const void *const fatlink,
+                  const void *const longlink, void *mg_pack_ptr, int mg_rebuild_type, void *source, void *solution,
+                  double *const final_residual, double *const final_fermilab_residual, int *num_iters)
+{
+  static const QudaVerbosity verbosity = getVerbosity();
+  qudamilc_called<true>(__func__, verbosity);
+
+  // FIXME: Flip the sign of the mass to fix a consistency issue between
+  // MILC, QUDA full parity dslash operator
+  mass = -mass;
+
+  milcMultigridPack *mg_pack = (milcMultigridPack *)(mg_pack_ptr);
+
+  if (target_fermilab_residual == 0 && target_residual == 0) errorQuda("qudaInvert: requesting zero residual\n");
+
+  // static const QudaVerbosity verbosity = getVerbosity();
+  QudaPrecision host_precision = (external_precision == 2) ? QUDA_DOUBLE_PRECISION : QUDA_SINGLE_PRECISION;
+  QudaPrecision device_precision = (quda_precision == 2) ? QUDA_DOUBLE_PRECISION : QUDA_SINGLE_PRECISION;
+  QudaPrecision device_precision_sloppy = QUDA_SINGLE_PRECISION; // required for MG
+
+  QudaGaugeParam fat_param = newQudaGaugeParam();
+  QudaGaugeParam long_param = newQudaGaugeParam();
+  setGaugeParams(fat_param, long_param, fatlink, longlink, localDim, host_precision, device_precision,
+                 device_precision_sloppy, inv_args.tadpole, inv_args.naik_epsilon);
+
+  fat_param.cuda_prec_refinement_sloppy = fat_param.cuda_prec_sloppy;
+  fat_param.cuda_prec_precondition = QUDA_HALF_PRECISION; // override
+  fat_param.reconstruct_refinement_sloppy = QUDA_RECONSTRUCT_NO;
+
+  long_param.type = QUDA_ASQTAD_LONG_LINKS;
+  long_param.cuda_prec_refinement_sloppy = long_param.cuda_prec_sloppy;
+  long_param.cuda_prec_precondition = QUDA_HALF_PRECISION; // override
+  long_param.reconstruct_refinement_sloppy = QUDA_RECONSTRUCT_NO;
+
+  QudaInvertParam invertParam = newQudaInvertParam();
+
+  QudaParity local_parity = inv_args.evenodd; // ignored, just needed to set some defaults
+  const double reliable_delta = 1e-4;
+
+  setInvertParams(localDim, host_precision, device_precision, device_precision_sloppy, mass, target_residual,
+                  target_fermilab_residual, inv_args.max_iter, reliable_delta, local_parity, verbosity,
+                  QUDA_GCR_INVERTER, &invertParam);
+
+  // Override a few things for WAR
+#if 0
```

Comment:
What are these lines about?  What's the WAR needed for?

Reply:
That can be cleaned up. This was from the earlier days of interface checking where I just wanted to make sure I was setting up full parity solves correctly, independent of what the solver was.